### ISSUE TITLE
is this memory leak?

### ISSUE THREAD

**Issue by Darren** [pingcap/tidb#7159]
In issue #7089 we met tidb-server OOM, I tried to find a recipe to reproduce, here's what I got:

1. start tidb-server with mocktikv
2. create table like this:
```sql
             CREATE TABLE if not exists t1 (
               id int(11) NOT NULL,
               kid bigint(20) UNSIGNED DEFAULT NULL,
               data blob DEFAULT NULL,
               nread int(11) DEFAULT NULL,
               nwrite int(11) DEFAULT NULL,
               PRIMARY KEY (id),
               KEY kid (kid)
           )
```
3. popluate the table with 500000 rows of sample data
4. update the table parallelly
5. watch tidb-server memory usage when update starts
6. we test on a server with 32CPUs and 128G of ram, the memory usage:
```
top - 00:58:29 up 6 days,  6:15,  1 user,  load average: 3.63, 3.75, 3.47
Tasks: 661 total,   1 running, 660 sleeping,   0 stopped,   0 zombie
Cpu(s):  5.1%us,  0.8%sy,  0.0%ni, 94.1%id,  0.0%wa,  0.0%hi,  0.0%si,  0.0%st
Mem:  132118812k total, 127355488k used,  4763324k free,   297536k buffers
Swap: 32768252k total,        0k used, 32768252k free, 26876568k cached

  PID USER      PR  NI  VIRT  RES  SHR S %CPU %MEM    TIME+  COMMAND                                                                                                                 
52347 root      20   0 68.0g  67g  13m S 159.3 54.0 421:21.54 tidb-server                   
```
7. tidb version:
```
Release Version: v2.0.5
Git Commit Hash: 43ab80012ae6a154e221e124a2ad3a396f0de822
Git Branch: release-2.0
UTC Build Time: 2018-07-09 01:30:21
GoVersion: go version go1.10.3 linux/amd64
TiKV Min Version: 2.0.0-rc.4.1
```

8. pprof heap top

```
(pprof) top
Showing nodes accounting for 30792.92MB, 99.21% of 31037.10MB total
Dropped 225 nodes (cum <= 155.19MB)
Showing top 10 nodes out of 42
      flat  flat%   sum%        cum   cum%
30163.37MB 97.18% 97.18% 30163.37MB 97.18%  github.com/pingcap/tidb/store/mockstore/mocktikv.appendRow (inline)
  623.55MB  2.01% 99.19%   623.55MB  2.01%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/util.(*BufferPool).Get
       5MB 0.016% 99.21%   628.55MB  2.03%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/table.(*Reader).readBlock
       1MB 0.0032% 99.21% 30238.87MB 97.43%  github.com/pingcap/tidb/store/mockstore/mocktikv.(*rpcHandler).handleCopDAGRequest
         0     0% 99.21%   186.88MB   0.6%  github.com/pingcap/tidb/executor.(*ExecStmt).Exec
         0     0% 99.21%   183.88MB  0.59%  github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor
         0     0% 99.21%   183.88MB  0.59%  github.com/pingcap/tidb/executor.(*UpdateExec).Next
         0     0% 99.21%   199.59MB  0.64%  github.com/pingcap/tidb/server.(*Server).onConn
         0     0% 99.21%   187.38MB   0.6%  github.com/pingcap/tidb/server.(*TiDBStatement).Execute
         0     0% 99.21%   187.88MB  0.61%  github.com/pingcap/tidb/server.(*clientConn).Run
```
9. the test programm itself

```go
package main

import (
	"crypto/rand"
	"database/sql"
	"flag"
	"fmt"
	"log"
	"sync"
	"time"

	_ "github.com/go-sql-driver/mysql"
)

var (
	host     = flag.String("host", "127.0.0.1", "tidb server address")
	port     = flag.Int("port", 4000, "tidb server port")
	limit    = flag.Int("limit", 500000, "how many rows to insert into the sample table")
	parallel = flag.Int("p", 64, "how many goroutines to start for updating the table")
	action   = flag.String("a", "", "action to execute: (insert|update)")

	initSQL = ` 
             CREATE TABLE if not exists t1 (
               id int(11) NOT NULL,
               kid bigint(20) UNSIGNED DEFAULT NULL,
               data blob DEFAULT NULL,
               nread int(11) DEFAULT NULL,
               nwrite int(11) DEFAULT NULL,
               PRIMARY KEY (id),
               KEY kid (kid)
             )`
)

// Executor the actioner
type Executor struct {
	dsn string
	db  *sql.DB
}

// Init Prepare the table
func (e *Executor) Init(host string, port int) (err error) {
	e.dsn = fmt.Sprintf("root:@tcp(%s:%d)/test", host, port)
	log.Printf("Using %s", e.dsn)
	e.db, err = sql.Open("mysql", e.dsn)
	return
}

// Create creates the table
func (e *Executor) Create() (err error) {
	log.Println(initSQL)
	_, err = e.db.Exec(initSQL)
	return
}

// Show the variables
func (e *Executor) Show() {
	rows, err := e.db.Query("show variables like '%commit%'")
	if err != nil {
		log.Println(err)
		return
	}
	defer rows.Close()

	var name, value string
	for rows.Next() {
		rows.Scan(&name, &value)
		log.Printf("%s -> %s", name, value)
	}
	return
}

func randBytes() []byte {
	b := make([]byte, 512)
	rand.Read(b)
	return b
}

func (e *Executor) insert(start, end int) error {
	tstart := time.Now()
	stmt, err := e.db.Prepare("replace into t1 (id, kid, data, nread, nwrite) values (?, ?, ?, ?, ?)")
	if err != nil {
		return err
	}
	defer stmt.Close()

	now := time.Now()
	total := float64(end - start)
	count := float64(0)

	for i := start; i < end; i++ {
		count++
		_, err := stmt.Exec(i, uint64(^0-i), randBytes(), i, i)
		if err != nil {
			return err
		}

		if i%1000 == 999 {
			log.Printf(
				"[Insert %06d-%d] complete:%d (%.2f%%), taken: %s",
				start, end, i-start,
				100*count/total,
				time.Since(now),
			)
			now = time.Now()
		}
	}

	log.Printf("Insert [%d to %d), time taken: %v", start, end, time.Since(tstart))
	return nil
}

// Insert populate the table with sample data
func (e *Executor) Insert() (err error) {
	var wg sync.WaitGroup

	batch := *limit / *parallel
	res := *limit % *parallel

	batchInsert := func(start, end int) {
		err := e.insert(start, end)
		if err != nil {
			log.Printf("Insert from %d  to %d failed, %v", start, end, err)
		}
		wg.Done()
	}

	i := 0
	for ; i < *parallel; i++ {
		wg.Add(1)
		go batchInsert(i*batch, (i+1)*batch)
	}

	if res > 0 {
		wg.Add(1)
		go batchInsert(i*batch, *limit)
	}

	wg.Wait()

	return nil
}

func (e *Executor) update(start, end int) error {
	tstart := time.Now()
	stmt, err := e.db.Prepare("update t1 set nread=nread+1, nwrite=nwrite + 1 where kid = ?")
	if err != nil {
		return err
	}
	defer stmt.Close()

	now := time.Now()
	total := float64(end - start)
	count := float64(0)

	for i := start; i < end; i++ {
		count++
		_, err := stmt.Exec(uint64(^0 - i))
		if err != nil {
			return err
		}

		if i%100 == 99 {
			log.Printf(
				"[update %06d-%d] complete:%d (%.2f%%), taken: %s",
				start, end, i-start,
				100*count/total,
				time.Since(now),
			)
			now = time.Now()
		}
	}

	log.Printf("Update [%d to %d), time taken: %v", start, end, time.Since(tstart))
	return nil
}

// Update update the table randomly
func (e *Executor) Update() (err error) {
	var wg sync.WaitGroup

	batch := *limit / *parallel
	res := *limit % *parallel

	batchUpdate := func(start, end int) {
		err := e.update(start, end)
		if err != nil {
			log.Printf("Update from %d  to %d failed, %v", start, end, err)
		}
		wg.Done()
	}

	i := 0
	for ; i < *parallel; i++ {
		wg.Add(1)
		go batchUpdate(i*batch, (i+1)*batch)
	}

	if res > 0 {
		wg.Add(1)
		go batchUpdate(i*batch, *limit)
	}

	wg.Wait()

	return nil
}

// Close shutdown the executor
func (e *Executor) Close() (err error) {
	return e.db.Close()
}

func main() {
	flag.Parse()

	var e Executor
	err := e.Init(*host, *port)
	if err != nil {
		log.Fatal(err)
	}

	e.Show()

	err = e.Create()
	if err != nil {
		log.Fatal(err)
	}

	switch *action {
	case "insert":
		err = e.Insert()
	case "update":
		err = e.Update()
	default:
		log.Printf("Unknown action: %s, start insert and update", *action)
		err = e.Insert()
		err = e.Update()
	}

	if err != nil {
		log.Fatal(err)
	}
}
```

**Comment by tiancaiamao** [comment]
Maybe the transaction never commit so TiDB hold all the data in the memory?
I'm not sure about the 'auto commit' session variable, but you can try to add some 'commit' statement during the batch insert.

If you start the binary with './tidb-server', take a look at the directory data size in '/tmp/tidb'.

**Comment by Darren** [comment]
> If you start the binary with './tidb-server', take a look at the directory data size in '/tmp/tidb'.

the /tmp/tidb is small
```sh
$ du -sch /tmp/tidb/
341M    /tmp/tidb/
341M    total
```

**Comment by tiancaiamao** [comment]
I've reproduce it, thanks for your report. @Darren

**Comment by Darren** [comment]
I have tested it against latest tidb-server (b5f9b35a453a0b5f06fd4a7fa77b94f251e7716e), 

the mem usage is much stabler, with 64 goroutines updating, the memory usage stop growing after it reaches 16GB.

I think the problem may have already been fixed.

**Comment by Darren** [comment]
@tiancaiamao 

After #7108 is merged, it is not reproducible with tidb-server tip.

I just want to know why tidb-server memory usage grows out of boundary if many expensive SQLs are being executed on the server.

Is it still possible that some bad written SQLs (like full table scan) in the application code cause tidb-server OOM again?

**Comment by tiancaiamao** [comment]
Hi, @Darren 

> Since #7108 is merged, it is not reproducible with tidb-server tip.

Good catch!

> I just want to know why tidb-server memory usage grows out of boundary if many expensive SQLs are being executed on the server.

I'm still investigating it, we need to find why the OOM happen before reaching a conclusion.
B.T.W, the SQL in the test code is not expensive query.

> Is it still possible that some bad written SQLs (like full table scan) in the application code cause tidb-server OOM again?

Probably. full table scan and some heavy hash join operation will cause tidb-server OOM.
We have some session variable to control the concurrency, and some session variable to limit a session's memory usage. 

In the latest tidb commit https://github.com/pingcap/tidb/pull/7108, I suspect the memory occupation is still abnormal.
When I find some clues, I'll post here. 
Please keep an eye on, thanks!

**Comment by tiancaiamao** [comment]
Before https://github.com/pingcap/tidb/pull/7108, the OOM is caused by the high concurrent full table scan, and related things are already explained in issue https://github.com/pingcap/tidb/issues/7089

Then, after running the test, I found tidb-server process will still take about 600M memory on my machine, and I suspected maybe there are leaks somewhere.

Now I confirmed that it's caused by mocktikv cache, which may still refer the allocated objects. 
Caches for performance is reasonable, those are not leaks.
If I using tidb-server with store=tikv, the tidb-server process will take very small memory footprint.

@Darren
